### PR TITLE
[14.0][ADD] sale_advance_payment

### PR DESCRIPTION
**Pull request by MateuGForgeFlow**
Allows the user to register payments in advance.

### CODE REVIEW

**Comment by JordiBForgeFlow** [comment]
@omar7r I hope that you're good that we migrate your module to OCA!

**Comment by mileo** [comment]
Nice module. 

I get this error trying to pay a S.O twice with the same journal.

image

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            payment_res = {
+                "payment_type": "inbound",
+                "partner_id": partner_id,
+                "partner_type": "customer",
+                "journal_id": self.journal_id.id,
+                "company_id": company.id,
+                "currency_id": sale.pricelist_id.currency_id.id,
+                "amount": self.amount_advance,
+                "sale_id": sale.id,
+                "name": _("Advance Payment") + " - " + sale.name,
```
Comment (mileo) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py]: Please check my comment and add a sequence here.

**Review by mileo — CHANGES_REQUESTED** [review]

**Comment by mileo** [comment]
Blocked by https://github.com/OCA/sale-workflow/pull/1478

https://travis-ci.com/github/OCA/sale-workflow/jobs/492812269#L1515

**Review by LoisRForgeFlow — APPROVED** [review]
LGTM

File: sale_advance_payment/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    def _compute_get_amount_residual(self):
+        for sale in self:
+            advance_amount = 0.0
+            for line in sale.account_payment_ids:
+                if line.state != "draft":
+                    advance_amount += line.amount
+            sale.amount_residual = sale.amount_total - advance_amount
+
+    account_payment_ids = fields.One2many(
```
Comment (rousseldenis) [inline sale_advance_payment/models/sale.py:12]: Could you put fields definition on top ?

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    _description = "Account Voucher Wizard"
+
+    @api.depends("journal_id")
+    def _compute_get_journal_currency(self):
+        for wzd in self:
+            wzd.journal_currency_id = (
+                wzd.journal_id.currency_id.id or self.env.user.company_id.currency_id.id
+            )
+
+    journal_id = fields.Many2one(
```
Comment (rousseldenis) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py:14]: Same as above. Put fields on top

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+        return res
+
+    @api.onchange("journal_id", "date", "amount_advance")
+    def onchange_date(self):
+
+        sale_ids = self.env.context.get("active_ids", [])
+        sale_obj = self.env["sale.order"]
+        if sale_ids:
+            sale_id = sale_ids[0]
```
Comment (rousseldenis) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py]: Please prefer fields.first()

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    def make_advance_payment(self):
+        """Create customer paylines and validates the payment"""
+        self.ensure_one()
+        payment_obj = self.env["account.payment"]
+        sale_obj = self.env["sale.order"]
+
+        sale_ids = self.env.context.get("active_ids", [])
+        if sale_ids:
+            sale_id = sale_ids[0]
```
Comment (rousseldenis) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py]: Same as above

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            sale = sale_obj.browse(sale_id)
+
+            partner_id = sale.partner_id.commercial_partner_id.id
```
Comment (rousseldenis) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py]: I'm not sure that's correct for standard flow. Should be at least partner_invoice_id no ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some remarks

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py (hunk 6 above)
Comment (MateuGForgeFlow) [reply sale_advance_payment/wizard/sale_advance_payment_wzd.py]: @rousseldenis You were right! Thank you for the review!

**Review by HaraldPanten — CHANGES_REQUESTED** [review]
As it is a new module, I was wondering if you could add a short usage explanation in the README file 👍

**Review by HaraldPanten — APPROVED** [review]
Functional review. Nice module 👍

File: sale_advance_payment/models/sale.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        ],
+        string="Advance Payment Status",
+        store=True,
+        readonly=True,
+        copy=False,
+        tracking=True,
+        compute="_compute_advance_payment",
+    )
+
+    @api.depends(
```
Comment (rousseldenis) [inline sale_advance_payment/models/sale.py:41]: Maybe a good improvement idea here is to split several computed values in separate fields (mls, advance_amount, amount_residual).

That allow a better comprehension of logic, and a better inheritance possibility.

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
+# Copyright (C) 2021 ForgeFlow S.L.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html)
+
+import json
+
+from odoo.exceptions import ValidationError
+from odoo.tests import common
+
+
+class TestRepairStockMove(common.SavepointCase):
```
Comment (rousseldenis) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: Could you change this ?

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        )
+        cls.product_2 = cls.env["product.product"].create(
+            {"name": "Conference Chair", "type": "product", "invoice_policy": "order"}
+        )
+        cls.product_3 = cls.env["product.product"].create(
+            {"name": "Repair Services", "type": "service", "invoice_policy": "order"}
+        )
+
+        # Location
+        cls.stock_warehouse = cls.env["stock.warehouse"].search(
```
Comment (rousseldenis) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: Your module don't depend on sale_stock. Are stock stuff testing needed ?

File: sale_advance_payment/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2015 Omar Castiñeira, Comunitea Servicios Tecnológicos S.L.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Sale Advance Payment",
+    "version": "14.0.1.0.0",
+    "author": "Comunitea, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/sale-workflow",
+    "category": "Sales",
+    "license": "AGPL-3",  # TODO: check
```
Comment (rousseldenis) [inline sale_advance_payment/__manifest__.py]: What's the status of this as mixed licenses through files ?

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
+##############################################################################
```
Comment (rousseldenis) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py]: Please change to short license version

File: sale_advance_payment/wizard/sale_advance_payment_wzd.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        payment_obj = self.env["account.payment"]
+        sale_obj = self.env["sale.order"]
+
+        sale_ids = self.env.context.get("active_ids", [])
+        if sale_ids:
+            sale_id = fields.first(sale_ids)
+            sale = sale_obj.browse(sale_id)
+
+            partner_id = sale.partner_invoice_id.commercial_partner_id.id
+            payment_res = {
```
Comment (rousseldenis) [inline sale_advance_payment/wizard/sale_advance_payment_wzd.py]: Maybe a _prepare_payment() ?

**Review by rousseldenis — COMMENTED** [review]
LGTM. Some improvement ideas.

**Comment by LoisRForgeFlow** [comment]
@mileo @rousseldenis Is this ready for you now? Could you update your review?

**Review by mileo — APPROVED** [review]
LGTM

**Comment by LoisRForgeFlow** [comment]
@rousseldenis good to merge?

**Comment by HaraldPanten** [comment]
@rousseldenis WDYT?

File: sale_advance_payment/models/sale.py (hunk 7 above)
Comment (rousseldenis) [reply sale_advance_payment/models/sale.py:41]: @MateuGForgeFlow Could you put a TODO here or a ROADMAP for this ?

Comment (MateuGForgeFlow) [reply sale_advance_payment/models/sale.py:41]: @rousseldenis Done! Added a ROADMAP. Thanks!

File: sale_advance_payment/views/sale_view.xml [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+<odoo>
+    <record id="view_order_form" model="ir.ui.view">
+        <field name="name">sale.order.form</field>
+        <field name="model">sale.order</field>
+        <field name="inherit_id" ref="sale.view_order_form" />
+        <field name="arch" type="xml">
+            <button name="action_draft" position="after">
+                <button
+                    name="%(action_view_account_voucher_wizard)d"
+                    string="Pay sale advanced"
```
Comment (MeetKD) [inline sale_advance_payment/views/sale_view.xml:11]: I prefer Sale advance payment or may be  Register Advance  Sale Payment. It just a suggestion.

**Review by MeetKD — APPROVED** [review]
There is one suggestion to change the button name but its totally upto you. nice module.

LGTM

**Review by JordiBForgeFlow — APPROVED** [review]

**Pull request merged**